### ISSUE TITLE
APPX Static Analysis Failure

### ISSUE THREAD

**Issue by ajinabraham** [DominikSchlecht/Mobile-Security-Framework-MobSF#10]
Environment Details
MobSF Host: OS OS X El Capitan
Windows VM : Windows 7 SP1, .NET 6.4, 64bit
MobSF Version: v0.9.2 Beta, Branch: windows_static

Error from MobSF Console

```
[INFO] MIME Type: application/octet-stream FILE: Client-for-Google-Drive.appx

[INFO] Performing Static Analysis of Windows APP
[30/Sep/2016 18:24:08] "POST /Upload/ HTTP/1.1" 200 160
[INFO] Windows Static Analysis Started
[INFO] Windows Binary Analysis Started
[INFO] Generating Hashes
[INFO] Extracting APPX
[INFO] Unzipping
[INFO] Starting Binary Analysis - XML
[INFO] Reading AppxManifest
[INFO] Starting Binary Analysis
[INFO] Windows VM configured.
[INFO] Uploading sample.
[INFO] Running binskim.
[WARNING] binskim has no results.
[INFO] Running binscope. This might take a while, depending on the binary size.

[2016-09-30 18:24:10]
[ERROR] Static Analyzer Windows (/Users/../Code/Mobile-Security-Framework-MobSF/StaticAnalyzer/views/windows.py, LINE 99 "bin_an_dic = _binary_analysis(app_dic['tools_dir'], app_dic['app_dir'])"): <Fault 1: "<class 'FileNotFoundError'>:[WinError 2] The system cannot find the file specified">
[30/Sep/2016 18:24:10] "GET /StaticAnalyzer_Windows/?name=Client-for-Google-Drive.appx&type=appx&checksum=496798cb0c0d1c084dd89dc5a5a1ada0 HTTP/1.1" 200 4637


```

From RPC Client

```
[*] Challenge successfully verified.
Analyzing...

Analysis finished but was not complete due to non-fatal runtime errors.
Unexpected runtime condition(s) observed: RuleNotApplicableToTarget
```

Other observations:
The files are uploaded to VM successfully. I can see the files under `Samples` directory

```
7dbeefc4f4885de8cd09e6200c50de8d
7dbeefc4f4885de8cd09e6200c50de8d_binskim
```

Tried with a different APPX file, same error

**Comment by ajinabraham** [comment]
Further Observations:

The `C:\MobSF\Tools\BinScope` directory is empty.

Also not sure if this is related: https://github.com/Microsoft/binskim/issues/39

**Comment by ajinabraham** [comment]
Reproduced the same issue in Windows 10 Home 64bit as well.

**Comment by sn0b4ll** [comment]
Hey Ajin,

first for BinScope:
setup.py states in in [line 228 to L232](https://github.com/DominikSchlecht/Mobile-Security-Framework-MobSF/blob/master/install/windows/setup.py#L228-L232) that BinScope has to be installed to that directory ^^ I will have to investigate if there's a way to automate this part.

To BinSkim:
BinSkim sometimes fails, when things like PEB is missing or the binary got "obfuscated" or compiled with flags which should prevent reverse-engineering. So if BinSkim works, the results should be good. If BinSkim fails, BinScope should do the job, that's why both of the are included :)
I will have to look at the policies for BinSkim, maybe I can design one which is more resistant to this kind of failure.

tl;dr: BinScope has to be installed to the directory by hand (sorry for that) and BinSkim is sometimes a bit touchy :)

**Comment by ajinabraham** [comment]
I just tried to configure it as any normal user would do. I think we need to add more info on how to configure this or perhaps simplify the process.

**Comment by ajinabraham** [comment]
> first for BinScope:
> setup.py states in in line 228 to L232 that BinScope has to be installed to that directory ^^ I will 
> have to investigate if there's a way to automate this part.

I had installed BinScope manually

> tl;dr: BinScope has to be installed to the directory by hand 

I had installed it, following the default install instructions. Do I have to install it to some specific directory like `C:\MobSF\Tools\BinScope`

**Comment by sn0b4ll** [comment]
Yepp, to `C:\MobSF\Tools\BinScope`. setup.py should also state this:
<img width="526" alt="screen shot 2016-10-02 at 00 32 40" src="https://cloud.githubusercontent.com/assets/1722036/19017391/e399c024-8837-11e6-8c43-628bab25fc34.png">

But I see that many users will miss this. I will have a look into it tomorrow.

**Comment by sn0b4ll** [comment]
### Shipping with MobSF

Sadly the EULA clearly states:

> SCOPE OF LICENSE. The software is licensed, not sold. This agreement only gives you some rights to use the software. Microsoft reserves all other rights. Unless applicable law gives you more rights despite this limitation, you may use the software only as expressly permitted in this agreement. In doing so, you must comply with any technical limitations in the software that only allow you to use it in certain ways. **You may not**
> - work around any technical limitations in the software;
> - reverse engineer, decompile or disassemble the software, except and only to the extent that applicable law expressly permits, despite this limitation;
> - make more copies of the software than specified in this agreement or allowed by applicable law, despite this limitation;
> - **publish the software for others to copy**;
> - rent, lease or lend the software; or
> - use the software for commercial software hosting services.

So we are not allowed to ship it with MobSF thank's to Microsoft strange EULAs...

### Testing for automatic installer
- Passive execution is working, but wrong folder is taken
  - Already Tried
    - msiexec /passive TARGETDIR="C:\MyTagretDirectory" /i MyProject.msi
    - msiexec /passive APPDIR="C:\MyTagretDirectory" /i MyProject.msi
    - msiexec /passive INSTALLPATH="C:\MyTagretDirectory" /i MyProject.msi
    - msiexec /passive INSTALLDIR="C:\MyTagretDirectory" /i MyProject.msi
    - msiexec /passive INSTALLFOLDER="C:\MyTagretDirectory" /i MyProject.msi

Using `msiexec /i BinScope_x64.msi /L*v everything.log` I was able to find the correct parameter `INSTALLLOCATION`, so with 
`msiexec INSTALLLOCATION="C:\MobSF\Tools\BinScope" /i BinScope_x64.msi /passive` passive and unattended installation is possible.

### Automatic download

Since we are not allowed to ship BinScope with MobSF, we must download it somehow.

I was able to sniff the following request for BinScope_x64.msi:
`https://download.microsoft.com/download/5/6/0/56027BE0-3A1F-4326-B966-BC99A15856AA/BinScope_x64.msi`
I don't know if the numbers in the middle are only from the cdn or have a session (with a timeout or smth..). What I can already say is that no cookies are required, since I was able to download BinScope from my virtual machine, capturing the link from my Mac-Host-System.

I will lay low for some hours (maybe hitting a timeout) and then try again. If this is working, I will implement the automatic installation for BinScope tonight or tomorrow.

**Comment by sn0b4ll** [comment]
Since the link is still working I will start to implement the automatic installer now.

**Comment by sn0b4ll** [comment]
Implemented with aad65a84860da4cd5e18db747437dfe6773c7bcb.

TODO: Check if the link stays online..

**Comment by ajinabraham** [comment]
My bad I missed the point, install to `C:\MobSF\Tools\BinScope`

**Comment by ajinabraham** [comment]
@DominikSchlecht I hope the download link stays like that.
How about adding a check/test at the end for all the required tools. Because most time, people will miss things like I did and open them as a bug. It will be difficult for us to track and we will get overwhelmed with unnecessary bugs. I am trying to do the same thing with other static analysis modules that requires third party module/software.

**Comment by ajinabraham** [comment]
@DominikSchlecht Now I had installed BinScope to `C:\MobSF\Tools\BinScope` but got a new error.

```
[INFO] Running binskim.

[2016-10-03 17:46:01]
[ERROR] Static Analyzer Windows (/Users/ajin/Code/Mobile-Security-Framework-MobSF/StaticAnalyzer/views/windows.py, LINE 99 "bin_an_dic = _binary_analysis(app_dic['tools_dir'], app_dic['app_dir'])"): <Fault 1: "<class 'KeyError'>:'file_x64'">
[03/Oct/2016 17:46:01] "GET /StaticAnalyzer_Windows/?name=Pocket-Explorer-1.1.1.62-Appx4Fun.com.appx&type=appx&checksum=788145f94013d3a41e8c0468c646e2d6 HTTP/1.1" 200 4637

```

FYI: I am using a 32 bit version of BinScope.

**Comment by sn0b4ll** [comment]
Will have a look into this tomorrow.

**Comment by sn0b4ll** [comment]
No wonder this is not working ^^
https://github.com/DominikSchlecht/Mobile-Security-Framework-MobSF/blob/master/StaticAnalyzer/views/windows.py#L286

Looking into it.

**Comment by sn0b4ll** [comment]
Will use `platform.machine().endswith('64')` for detecting x64 hosts.

**Comment by sn0b4ll** [comment]
Hopefully fixed with 7980b6a.

Could you please verify? Will issue an PR (after merging with upstream).

**Comment by ajinabraham** [comment]
All Good. Works Fine!
